File: 49주차/두큐합같게만들기.py
from collections import deque
def solution(q1,q2):
    answer = 0
    q1 = deque(q1)
    q2 = deque(q2)
    cycle = (len(q1) + len(q2))*2
    
    q1Total = sum(q1)
    q2Total = sum(q2)
    target = q1Total + q2Total
    if target % 2 !=0:
        return -1
    target = target//2
    
    while True:
        if cycle < answer:
            return -1
        if q1Total == q2Total:
            return answer
        if q1Total > target:
            v = q1.popleft()
            q2.append(v)
            q1Total -= v
            q2Total += v
        else:
            v = q2.popleft()
            q1.append(v)
            q2Total -= v
            q1Total += v
        answer += 1
# 큐에서 직접 pop, insert해서 풀기.

File: 49주차/두큐합같게만들기.js
function solution(queue1, queue2) {
  var answer = 0;
  const arr = [...queue1, ...queue2];
  const sum = (arr) => arr.reduce((a, b) => a + b); // 전체 합을 구하는 함수.
  let start = 0;
  let end = queue1.length;
  let total = sum(arr.slice(start, end));
  const target = sum(arr) / 2; // 만들고자하는 합.

  const maxCnt = queue1.length * 3; // start는 2*n만큼, end는 n만큼 이동가능하므로 총 3*n 만큼이동.
  while (answer <= maxCnt) {
    if (total < target) {
      // 타겟보다 작으면 end +1
      total += arr[end];
      end++;
    } else if (total === target) return answer;
    else {
      total -= arr[start]; // 타겟보다 크다면 start + 1
      start++;
    }
    answer += 1;
  }

  return -1; // 같게 만들어지지 않으면 -1.
}
// 프로그래머스 문제
// 투포인터를 이용해서 js로 구현.
